src/Share/BackgroundJobs/Diffs/ContributionDiffs.hs

```diff
@@ -23,52 +23,81 @@ import Share.Utils.Logging qualified as Logging
 import Share.Web.Authorization qualified as AuthZ
 import Share.Web.Errors (EntityMissing (..), ErrorID (..))
 import Share.Web.Share.Diffs.Impl qualified as Diffs
+import System.Clock qualified as Clock
 
--- | Check every 10 minutes if we haven't heard on the notifications channel.
+-- | Check every 30 seconds if we haven't heard on the notifications channel.
 -- Just in case we missed a notification.
 maxPollingIntervalSeconds :: Int
-maxPollingIntervalSeconds = 10 * 60
+maxPollingIntervalSeconds = 30
 
 worker :: Ki.Scope -> Background ()
 worker scope = do
   authZReceipt <- AuthZ.backgroundJobAuthZ
+  badUnliftCodebaseRuntime <- Codebase.badAskUnliftCodebaseRuntime
+  unisonRuntime <- asks Env.sandboxedRuntime
+  let makeRuntime :: Codebase.CodebaseEnv -> IO (Codebase.CodebaseRuntime IO)
+      makeRuntime codebase = do
+        runtime <- Codebase.codebaseRuntime' unisonRuntime codebase
+        pure (badUnliftCodebaseRuntime runtime)
   newWorker scope "diffs:contributions" $ forever do
     Notif.waitOnChannel Notif.ContributionDiffChannel (maxPollingIntervalSeconds * 1000000)
-    processDiffs authZReceipt >>= \case
-      Left (contributionId, e) ->
-        withTag "contribution-id" (IDs.toText contributionId) $ do
-          reportError e
-      Right _ -> pure ()
+    processDiffs authZReceipt makeRuntime
 
-processDiffs :: AuthZ.AuthZReceipt -> Background (Either (ContributionId, NamespaceDiffError) ())
-processDiffs authZReceipt = Metrics.recordContributionDiffDuration . runExceptT $ do
-  mayContributionId <- PG.runTransaction DQ.claimContributionToDiff
-  for_ mayContributionId (diffContribution authZReceipt)
-  case mayContributionId of
-    Just contributionId -> do
-      Logging.textLog ("Recomputed contribution diff: " <> tShow contributionId)
-        & Logging.withTag ("contribution-id", tShow contributionId)
-        & Logging.withSeverity Logging.Info
-        & Logging.logMsg
-      -- Keep processing releases until we run out of them.
-      either throwError pure =<< lift (processDiffs authZReceipt)
-    Nothing -> pure ()
+-- Process diffs until we run out of them. We claim a diff in a transaction, commit it, then proceed to compute the
+-- diff. There's therefore a chance we claim a diff and fail to compute it (due to e.g. server restart). The current
+-- solution to these "at most once" semantics is to simply re-enqueue a diff job if necessary; e.g. in the view diff
+-- endpoint handler.
+processDiffs :: AuthZ.AuthZReceipt -> (Codebase.CodebaseEnv -> IO (Codebase.CodebaseRuntime IO)) -> Background ()
+processDiffs authZReceipt makeRuntime = do
+  let loop :: Background ()
+      loop = do
+        result <-
+          PG.runTransactionMode PG.RepeatableRead PG.ReadWrite do
+            DQ.claimContributionToDiff >>= \case
+              Nothing -> pure Nothing
+              Just contributionId -> do
+                startTime <- PG.transactionUnsafeIO (Clock.getTime Clock.Monotonic)
```

Comment:
Unless you really think it's important that this capture _only_ the diff calculation you can probably just use `timeActionIntoHistogram` outside the transaction; unless you're worried that'll capture noise a'la pg waiting for locks or something?

Reply:
Oh, good catch, so – there was a time when this logic was intended only to capture timing metrics for successful diffs, and ignore diff errors (i.e. due to incoherent decl or whatever). I thought that would give us a more informative historgram.

But now, it doesn't really do that, since we compute & store a diff, failure or not.

Reply:
True